HW #6 (Monolith and microservices)

cholombytko opened the pull request:
This pull request contains refactoring process according to common architecture practices

cholombytko committed:
refactor: project structure refactor

This commit contains refactor of project structure by creating common
directory

cholombytko committed:
refactor: delete unused file

This commit contains refactor of project by deleting old version of
architecture diagram

cholombytko committed:
Revert "refactor: delete unused file"

This reverts commit 2659b07e24376b8add0724120a79bcbd97aef5f7.
The reason is that this commit should be related to another pull request

cholombytko committed:
feat: add rabbit mq to docker compose

This commit contains adding rabbit mq to docker compose file

cholombytko committed:
Merge branch 'hw7/microservice-patterns' of github.com:GenesisEducationKyiv/software-engineering-school-4-0-cholombytko into hw6/monolith

cholombytko committed:
feat: subsctiption microservice basic implementation

This commit contains basic implementation of subscription microservice

cholombytko committed:
feat: implementation of api-gateway, rate and subscription microservices

This commit contains several updates:
-First version of api gateway microservice
-Subscription service changes and updates
-Rate service implementation
-Updating docker-compose for service workflow

cholombytko committed:
feat: email-sender implementation

This commit contains basic email-sender microservice implementaion

eluyema commented on package.json:
```diff
@@ -44,6 +44,7 @@
   },
   "devDependencies": {
```
You can remove extra dependencies from this file, for example - typeorm, pg, nodemailer etc

cholombytko replied:
Great idea, thanks!

eluyema commented on rate/src/app.controller.ts:
```diff
@@ -0,0 +1,10 @@
+import { RATE_SERVICE_TOKEN } from './app.constants';
+
+@Controller()
+export class AppController {
```
Do you need this file?

eluyema replied:
This controller is empty, so which controller do you use for this microservice ?

cholombytko replied:
Yeah, i know. Several microservices don't have connection yet, because of @nestjs/microservices bug in mine project. I already have [BugReport](https://github.com/nestjs/nest/issues/13769) in @nestjs

eluyema commented on src/app.module.ts:
```diff
@@ -1,6 +1,6 @@
+import { DatabaseModule } from './common/database.module';
+import { MailModule } from './common/mail.module';
 import { EmailModule } from './email/email.module';
-import { MailModule } from './mail.module';
```
Do you steel need this src folder?
Maybe all responsibilities of this application are now handled by the Rate, Email-Sender, and Subscription services, aren't they?

cholombytko replied:
Yes, i left this folder for overview process. It will be deleted in next commit

eluyema commented on subscription/src/app.controller.ts:
```diff
@@ -0,0 +1,18 @@
+    private readonly subscriptionService: ISubscriptionService,
+  ) {}
+
+  @EventPattern('add-subscriber')
```
Should this route maybe be done in a synchronous way?

eluyema commented on subscription/src/main.ts:
```diff
@@ -0,0 +1,27 @@
+        },
+        urls: [`amqp://${RMQ_USER}:${RMQ_PASSWORD}@rabbitmq:5672`],
+      },
+      transport: Transport.RMQ,
```
It's great that you decided to use RabbitMQ

eluyema reviewed (changes requested):
Well done!
Just fix some places with extra code.

cholombytko committed:
feat: delete unused src folder

This commit contains deleting old src folder which was used in
monolithic arch

cholombytko committed:
feat: rate service connection implementation

This commit contains implementation of connection between api-gateway
and rate service using rmq

cholombytko committed:
fix: some minor changes for subscription service

cholombytko committed:
HW7 feat: email sender implementation

This commit contains email sender service implementation and rmq
connection to it

Larisa-Antsifrova commented on api-gateway/test/app.e2e-spec.ts:
```diff
@@ -0,0 +1,24 @@
+    return request(app.getHttpServer())
+      .get('/')
+      .expect(200)
+      .expect('Hello World!');
```
That is an interesting set of tests 😅

Larisa-Antsifrova commented on test/email/email.integration.spec.ts:
```diff
@@ -1,5 +1,5 @@
 /* eslint-disable @typescript-eslint/no-unused-vars */
```
I am curious why this set of tests is not located in any of the modules (emails, for example)?)

Larisa-Antsifrova commented on email-sender/src/main.ts:
```diff
@@ -0,0 +1,27 @@
+        },
+        urls: [`amqp://${RMQ_USER}:${RMQ_PASSWORD}@rabbitmq:5672`],
+      },
+      transport: Transport.RMQ,
```
Non-blocking: It seems that this config could have been placed somewhere else, and named accordingly :)

Larisa-Antsifrova commented on email-sender/src/main.ts:
```diff
@@ -0,0 +1,27 @@
+  );
+  await app.listen();
+}
+bootstrap();
```
Also, I might be missing it out, but I cannot see the mechanism of a timed email sending. If you could direct me to it, I would appreciate it :)
